server/src/main/java/org/apache/druid/discovery/DataServerClient.java

```diff
@@ -0,0 +1,171 @@
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.druid.discovery;
+
+import com.fasterxml.jackson.databind.JavaType;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import com.fasterxml.jackson.dataformat.smile.SmileFactory;
+import com.google.common.util.concurrent.FutureCallback;
+import com.google.common.util.concurrent.Futures;
+import com.google.common.util.concurrent.ListenableFuture;
+import org.apache.druid.client.JsonParserIterator;
+import org.apache.druid.java.util.common.concurrent.Execs;
+import org.apache.druid.java.util.common.guava.BaseSequence;
+import org.apache.druid.java.util.common.guava.Sequence;
+import org.apache.druid.java.util.common.io.Closer;
+import org.apache.druid.java.util.common.logger.Logger;
+import org.apache.druid.java.util.http.client.response.StatusResponseHandler;
+import org.apache.druid.java.util.http.client.response.StatusResponseHolder;
+import org.apache.druid.query.Query;
+import org.apache.druid.query.context.ResponseContext;
+import org.apache.druid.rpc.FixedSetServiceLocator;
+import org.apache.druid.rpc.RequestBuilder;
+import org.apache.druid.rpc.ServiceClient;
+import org.apache.druid.rpc.ServiceClientFactory;
+import org.apache.druid.rpc.StandardRetryPolicy;
+import org.apache.druid.utils.CloseableUtils;
+import org.jboss.netty.handler.codec.http.HttpMethod;
+
+import java.io.InputStream;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.Future;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.TimeUnit;
+
+/**
+ * Client to query data servers given a query.
+ */
+public class DataServerClient
+{
+  private static final Logger log = new Logger(DataServerClient.class);
+  private final ServiceClient serviceClient;
+  private final ObjectMapper objectMapper;
+  private final ScheduledExecutorService queryCancellationExecutor;
+
+  public DataServerClient(
+      ServiceClientFactory serviceClientFactory,
+      FixedSetServiceLocator fixedSetServiceLocator,
+      ObjectMapper objectMapper
+  )
+  {
+    this.serviceClient = serviceClientFactory.makeClient(
+        NodeRole.INDEXER.getJsonName(),
+        fixedSetServiceLocator,
+        StandardRetryPolicy.noRetries()
+    );
+    this.objectMapper = objectMapper;
+    this.queryCancellationExecutor = Execs.scheduledSingleThreaded("query-cancellation-executor");
```

Comment:
Can we make a pool for this. For each segment you are creating a new executor service which seems wasteful.

Reply:
Moved this to a common threadpool and added a wait to close it.